### PR TITLE
Pipe: support alter pipe functionality

### PR DESCRIPTION
**Pull request by VGalaxies**
## Description

As title, this PR introduces the Alter Pipe SQL statement for dynamically updating an existing pipe's processor and sink plugins. The general syntax is as follows:

```sql
alter pipe <pipe_name> modify processor (...) modify sink (...)
```

**NOTE:**

1. The key-value pairs of processor and sink plugins provided in the alter statement will override the previous plugin key-value pairs.

	For example, if the parameter `batch.enable` in the sink plugin is specified as false during the create pipe operation, but not specified during the alter pipe operation, then after the alter operation, the `batch.enable` parameter will not exist in the sink plugin of the pipe (default value is true).

2. Processor or sink plugins that are not modified in the alter statement will reuse the previous configuration.

	For example, if `alter pipe a2b modify processor (...)`, then the sink plugin of pipe `a2b` will reuse the previous configuration.

3. If the pipe's status was user stopped before the alter operation, then after the alter operation, the pipe's state remains user stopped. In all other cases, after the alter operation, the pipe's status becomes running, and the exception messages are cleared.

4. The creation time of the pipe will be updated after the alter operation.

5. The progress index of the pipe will be retained after the alter operation.

6. If the pipe corresponding to `<pipe_name>` does not exist, or if there is no change in the plugin key-value pairs before and after the alter pipe operation, the alter operation will fail.

<hr>

This PR has:
- [x] been self-reviewed.
    - [ ] concurrent read
    - [ ] concurrent write
    - [ ] concurrent read and write 
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. 
- [ ] added or updated version, __license__, or notice information
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [x] added integration tests.
- [x] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/write/pipe/task/AlterPipePlanV2.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import org.apache.iotdb.commons.pipe.task.meta.PipeRuntimeMeta;
+import org.apache.iotdb.commons.pipe.task.meta.PipeStaticMeta;
+import org.apache.iotdb.confignode.consensus.request.ConfigPhysicalPlan;
+import org.apache.iotdb.confignode.consensus.request.ConfigPhysicalPlanType;
+
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+public class AlterPipePlanV2 extends ConfigPhysicalPlan {
```
Comment (Caideyipi) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/write/pipe/task/AlterPipePlanV2.java:31]: Add a ser/de test in the "ConfigPhysicalPlanSerDeTest"

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/pipe/PipeTaskInfo.java [hunk 2]
```diff
@@ -158,6 +162,61 @@ private void checkBeforeCreatePipeInternal(TCreatePipeReq createPipeRequest)
...
+    acquireReadLock();
+    try {
+      checkBeforeAlterPipeInternal(alterPipeRequest);
+    } finally {
+      releaseReadLock();
+    }
+  }
+
+  private void checkBeforeAlterPipeInternal(TAlterPipeReq alterPipeRequest) throws PipeException {
+    if (isPipeExisted(alterPipeRequest.getPipeName())) {
```
Comment (SteveYurongSu) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/pipe/PipeTaskInfo.java:175]: fast failing here should be better

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/pipe/task/AlterPipeProcedureV2.java [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+  }
+
+  public AlterPipeProcedureV2(TAlterPipeReq alterPipeRequest) throws PipeException {
+    super();
+    this.alterPipeRequest = alterPipeRequest;
+  }
+
+  @Override
+  protected PipeTaskOperation getOperation() {
+    return PipeTaskOperation.CREATE_PIPE;
```
Comment (SteveYurongSu) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/pipe/task/AlterPipeProcedureV2.java:74]:
```suggestion
    return PipeTaskOperation.ALTER_PIPE;
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/execution/config/executor/ClusterConfigTaskExecutor.java [hunk 4]
```diff
@@ -1655,6 +1660,91 @@ public SettableFuture<ConfigTaskResult> createPipe(CreatePipeStatement createPip
     return future;
   }
 
+  @Override
+  public SettableFuture<ConfigTaskResult> alterPipe(AlterPipeStatement alterPipeStatement) {
+    SettableFuture<ConfigTaskResult> future = SettableFuture.create();
+
+    // Get pipe static meta
+    final String pipeName = alterPipeStatement.getPipeName();
+    final PipeStaticMeta pipeStaticMeta = PipeAgent.task().getPipeStaticMeta(pipeName);
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/execution/config/executor/ClusterConfigTaskExecutor.java:1669]: The pipeTaskAgent may not know this... Like when previous synchronizations are failed due to a dead address at connector: Consider this, a "createPipe" successfully handshakes with the receiver before configNode knows it, but the receiver is crushed just before the "CreatePipeProcedureV2" push the pipeMeta to dataNode. In this circumstance, the pipeTaskAgent does not know this pipe. You should allow users to alter a pipe like this that can not be created, and consequently won't be recorded at pipeTaskAgent.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/execution/config/executor/ClusterConfigTaskExecutor.java [hunk 5]
```diff
@@ -1655,6 +1660,91 @@ public SettableFuture<ConfigTaskResult> createPipe(CreatePipeStatement createPip
...
+    if (Objects.isNull(pipeStaticMeta)) {
+      future.setException(
+          new IoTDBException(
+              String.format("Failed to alter pipe %s, the pipe does not exist", pipeName),
+              TSStatusCode.PIPE_ERROR.getStatusCode()));
+      return future;
+    }
+
+    // We do not support alter source plugin of pipe, so the previous configuration will be reused.
+    alterPipeStatement.setExtractorAttributes(
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/execution/config/executor/ClusterConfigTaskExecutor.java:1679]: I suggest pass the order directly to configNode here without checking logic other than the checks in PipePluginAgent.

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/pipe/task/AlterPipeProcedureV2.java [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+  protected void rollbackFromWriteConfigNodeConsensus(ConfigNodeProcedureEnv env) {
+    LOGGER.info(
+        "AlterPipeProcedureV2: rollbackFromWriteConfigNodeConsensus({})",
+        alterPipeRequest.getPipeName());
+    TSStatus response;
+    try {
+      response =
+          env.getConfigManager()
+              .getConsensusManager()
+              .write(new DropPipePlanV2(alterPipeRequest.getPipeName()));
```
Comment (SteveYurongSu) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/pipe/task/AlterPipeProcedureV2.java:195]: ..

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/pipe/PipeTaskInfo.java [hunk 7]
```diff
@@ -283,6 +341,19 @@ public TSStatus createPipe(CreatePipePlanV2 plan) {
     }
   }
 
+  public TSStatus alterPipe(AlterPipePlanV2 plan) {
+    acquireWriteLock();
+    try {
+      pipeMetaKeeper.removePipeMeta(plan.getPipeStaticMeta().getPipeName());
+      pipeMetaKeeper.addPipeMeta(
+          plan.getPipeStaticMeta().getPipeName(),
+          new PipeMeta(plan.getPipeStaticMeta(), plan.getPipeRuntimeMeta()));
```
Comment (Caideyipi) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/pipe/PipeTaskInfo.java:350]: I strongly suggest that updating progressIndex also here... What about DNs are crushed? Then they restart and fetch only a MinimumProgressIndex at configNode...

**Review by Caideyipi — COMMENTED** [review]
PTAL

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/pipe/task/AlterPipeProcedureV2.java [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+    if (!exceptionMessage.isEmpty()) {
+      LOGGER.warn(
+          "Failed to rollback alter pipe {}, details: {}, metadata will be synchronized later.",
+          alterPipeRequest.getPipeName(),
+          exceptionMessage);
+    }
+  }
+
+  @Override
+  public void serialize(DataOutputStream stream) throws IOException {
```
Comment (SteveYurongSu) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/pipe/task/AlterPipeProcedureV2.java:224]: In theory, we need to serialize all members, including alterPipeRequest, pipeStaticMeta, and pipeRuntimeMeta

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/pipe/task/AlterPipeProcedureV2.java [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+              if (regionGroupId.getType().equals(TConsensusGroupType.DataRegion)) {
+                final String databaseName =
+                    env.getConfigManager()
+                        .getPartitionManager()
+                        .getRegionStorageGroup(regionGroupId);
+                if (databaseName != null && !databaseName.equals(SchemaConstant.SYSTEM_DATABASE)) {
+                  // Pipe only collect user's data, filter metric database here.
+                  consensusGroupIdToTaskMetaMap.put(
+                      regionGroupId,
+                      new PipeTaskMeta(MinimumProgressIndex.INSTANCE, regionLeaderNodeId));
```
Comment (SteveYurongSu) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/pipe/task/AlterPipeProcedureV2.java:124]: ..

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/pipe/task/AlterPipeProcedureV2.java [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+                }
+              }
+            });
+    pipeRuntimeMeta = new PipeRuntimeMeta(consensusGroupIdToTaskMetaMap);
+    pipeRuntimeMeta.getStatus().set(PipeStatus.RUNNING);
```
Comment (SteveYurongSu) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/pipe/task/AlterPipeProcedureV2.java:129]: USER STOPPED

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/pipe/PipeTaskInfo.java [hunk 11]
```diff
@@ -158,6 +162,55 @@ private void checkBeforeCreatePipeInternal(TCreatePipeReq createPipeRequest)
...
+        .equals(new TreeMap<>(alterPipeRequest.getProcessorAttributes()).toString())) {
+      needToAlter = true;
+    }
+    if (alterPipeRequest.getConnectorAttributes().isEmpty()) {
+      alterPipeRequest.setConnectorAttributes(
+          pipeStaticMetaFromCoordinator.getConnectorParameters().getAttribute());
+    } else if (!(new TreeMap<>(
+                pipeStaticMetaFromCoordinator.getConnectorParameters().getAttribute())
+            .toString())
+        .equals(new TreeMap<>(alterPipeRequest.getConnectorAttributes()).toString())) {
```
Comment (Caideyipi) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/pipe/PipeTaskInfo.java:202]: consider using keyReducer to detect identical contents like "source.xxx", "connector.xxx" or "xxx"

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/pipe/PipeTaskInfo.java [hunk 12]
```diff
@@ -259,11 +312,12 @@ public boolean isPipeRunning(String pipeName) {
     }
   }
 
-  public boolean isPipeStopped(String pipeName) {
+  public boolean isPipeUserStopped(String pipeName) {
```
Comment (Caideyipi) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/pipe/PipeTaskInfo.java:315]: is pipe stopped by user

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/pipe/task/AlterPipeProcedureV2.java [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+    if (updatedPipeRuntimeMeta != null) {
+      ReadWriteIOUtils.write(true, stream);
+      updatedPipeRuntimeMeta.serialize(stream);
+    } else {
+      ReadWriteIOUtils.write(false, stream);
+    }
+  }
+
+  @Override
+  public void deserialize(ByteBuffer byteBuffer) {
```
Comment (Caideyipi) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/pipe/task/AlterPipeProcedureV2.java:284]: consider adding a ser/de test as AlterPipeProcedureV2Test(can refer to CreatePipeProcedureV2Test)

Comment (Caideyipi) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/pipe/task/AlterPipeProcedureV2.java:284]: Consider adding a ser/de test as AlterPipeProcedureV2Test(can refer to CreatePipeProcedureV2Test)

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/pipe/agent/task/PipeTaskAgent.java [hunk 14]
```diff
@@ -797,8 +797,7 @@ protected boolean checkBeforeDropPipe(
    * @return {@code true} if need to drop pipe tasks, {@code false} if no need to drop.
    * @throws IllegalStateException if current pipe status is illegal.
```
Comment (Caideyipi) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/pipe/agent/task/PipeTaskAgent.java:798]: If do not throw exception also alter the @throws tag.

**Review by Caideyipi — CHANGES_REQUESTED** [review]
There are still some suggested changes :-)

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/pipe/PipeTaskInfo.java (hunk 11 above)
Comment (SteveYurongSu) [reply iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/pipe/PipeTaskInfo.java]: To make it simple, we can ignore these cases

**Review by Caideyipi — COMMENTED** [review]
Something about IT~

**Review by SteveYurongSu — APPROVED** [review]
What a great job!

**Pull request merged**